app/components/root/root.js

```diff
@@ -29,21 +40,98 @@ export default class Root extends React.Component {
     componentDidMount() {
         AppState.addEventListener('change', this.handleAppStateChange);
         EventEmitter.on(Constants.CONFIG_CHANGED, this.handleConfigChanged);
+        MessageBarManager.registerMessageBar(this.refs.notification);
+        this.configurePushNotifications();
     }
 
     componentWillUnmount() {
         AppState.removeEventListener('change', this.handleAppStateChange);
         EventEmitter.off(Constants.CONFIG_CHANGED, this.handleConfigChanged);
+        MessageBarManager.unregisterMessageBar();
+        PushNotification.unregister();
     }
 
+    onRegisterDevice = (data) => {
+        const prefix = Platform.OS === 'ios' ? Constants.PUSH_NOTIFY_APPLE_REACT_NATIVE : Constants.PUSH_NOTIFY_ANDROID_REACT_NATIVE;
+        this.props.actions.setDeviceToken(`${prefix}:${data.token}`);
+    };
+
+    onPushNotification = (notification) => {
+        const {foreground, userInteraction, data, message} = notification;
+        let noty;
+
+        if (Platform.OS === 'android') {
+            noty = {
+                data: {
+                    channel_id: notification.channel_id,
+                    team_id: notification.team_id
+                },
+                message
+            };
+        } else {
+            noty = {
+                data,
+                message
+            };
+        }
+
+        if (foreground) {
+            this.handleInAppNotification(noty);
+        } else if (userInteraction) {
+            this.onNotificationTapped(noty);
+        }
+    };
+
+    onNotificationTapped = (notification) => {
+        const {currentTeamId, currentChannelId} = this.props;
+
+        if (currentTeamId && currentChannelId) {
+            // this means that the store has the necessary data
+            this.props.actions.goToNotification(notification);
+        } else {
+            this.props.actions.queueNotification(notification);
+        }
+    };
+
     handleAppStateChange(appState) {
         this.props.actions.setAppState(appState === 'active');
+
+        if (appState === 'inactive') {
+            this.props.actions.flushToStorage();
+        }
     }
 
     handleConfigChanged = (serverVersion) => {
         this.props.actions.loadConfigAndLicense(serverVersion);
     };
 
+    handleInAppNotification = (notification) => {
+        const {message} = notification;
+
+        MessageBarManager.showAlert({
+            alertType: 'info',
+            avatar: icon,
+            avatarStyle: {borderRadius: 10, width: 20, height: 20},
+            message,
+            stylesheetInfo: {backgroundColor: '#9c9c9c'},
+            messageStyle: {color: 'white', fontSize: 13},
+            viewTopInset: 15,
+            viewBottomInset: 15,
+            duration: 5000,
+            onTapped: this.onNotificationTapped.bind(this, notification)
```

Comment:
actually I just remembered why, the `onTapped` function does not return anything, but I can change it to

```javascript
onTapped: () => this..onNotificationTapped(notification)